sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala

```diff
@@ -120,17 +120,19 @@ object RowEncoder {
 
     case StructType(fields) =>
       val convertedFields = fields.zipWithIndex.map { case (f, i) =>
-        val method = if (f.dataType.isInstanceOf[StructType]) {
-          "getStruct"
+        val x = extractorsFor(
+          GetExternalRowField(inputObject, i, externalDataTypeForInput(f.dataType), f.nullable),
+          f.dataType
+        )
+        if (f.nullable) {
+          If(
```

Comment:
i had that before (basically without the If, since GetExternalRowField already does null checks inside), but the issue is that extractorsFor code path then also runs for nulls , and that causes some weird runtime errors. i will try again.